### PR TITLE
[🦩] Add swiftASTGen library to build a C++ AST from a SwiftSyntax parse tree

### PR DESCRIPTION
**Pull request by DougGregor**
Introduce a `swiftASTGen` library, written as a Swift module, which produces a C++ AST from the corresponding SwiftSyntax syntax tree as produced by the new Swift parser. Architecturally, this new library sits between the C++ AST library (since it constructs ASTs) and the C++ parser (whose entry points will be diverted through this library). This new code path intercepts the main entry point of the C++ parser, which parses "top-level" declarations, when the command-line option `--enable-experimental-feature ParserASTGen`. It doesn't do much yet; this pull request is primarily to get all of the major architectural pieces in place so we can iterate on them.

`swiftASTGen` depends on the `CASTBridging` layer, a painstakingly-crafted C API that is used to build and inspect C++ AST nodes, so they can be used from Swift. We cannot currently depend on Swift/C++ interoperability, because it is an experimental feature, whereas this library will need to be able to build with a host Swift compiler from up to a year ago. 

The CMake build system is the only one that's able to produce a working compiler. There is also a `Package.swift` that makes it easier to build and work with the `swiftASTGen` library within IDEs, but it's mainly there for editing---it won't create something that can be meaningfully executed. Most things with the new Swift parser are better implemented/tested within or on top of the `swift-syntax` package.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by zoecarver** [comment]
Thank you for fixing this up, Doug!

**Review by zoecarver — APPROVED** [review]

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Review by etcwilde — COMMENTED** [review]
Lets remove the `xcworkspacedata` file and Zoe's `UserInterfaceState.xcuserstate`.

File: lib/ASTGen/Sources/ASTGen/ASTGen.swift [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+extension UnsafePointer {
+  public var raw: UnsafeMutableRawPointer {
+    UnsafeMutableRawPointer(mutating: self)
+  }
+}
+
+// TODO: :(
+var declContext: UnsafeMutableRawPointer! = nil
```
Comment (krzyzanowskim) [inline lib/ASTGen/Sources/ASTGen/ASTGen.swift:21]: mutable module-wide global variable buried between the lines.

It feels like it supposes to be a `ASTGenVisitor` property set in the `ASTGenVisitor` initializer. Unfortunate `ASTGenVisitor` is a struct that makes it all `mutating`. How about wrapping it in a class

```swift
class DeclContext {
    var declContext: UnsafeMutableRawPointer! = nil
}

struct ASTGenVisitor ... {
    var declContext = DeclContext()
}

```

and yes, nobody asked me for this comment

Comment (DougGregor) [reply lib/ASTGen/Sources/ASTGen/ASTGen.swift]: Gaaa, I hadn't noticed that. I've done a Sneaky Property Wrapper Trick for the moment.

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
This *really* doesn't do much now, and lots of stuff is likely to get reworked. We need to do a deep dive on a small part of the grammar (types and `typealias` is probably the smallest bit) and make things as clean and complete as possible, and see how that goes.

File: lib/AST/DeclContext.cpp [hunk 2]
```diff
@@ -882,7 +882,7 @@ void IterableDeclContext::addMember(Decl *member, Decl *hint, bool insertAtHead)
 
 void IterableDeclContext::addMemberSilently(Decl *member, Decl *hint,
                                             bool insertAtHead) const {
-  assert(!isa<AccessorDecl>(member) && "Accessors should not be added here");
+//  assert(!isa<AccessorDecl>(member) && "Accessors should not be added here");
```
Comment (ahoppen) [inline lib/AST/DeclContext.cpp:885]: Is the removal of this assertion intentional?

File: lib/ASTGen/Package.swift [hunk 3]
```diff
@@ -0,0 +1,32 @@
+// swift-tools-version: 5.6
+
+import PackageDescription
+
+let package = Package(
```
Comment (ahoppen) [inline lib/ASTGen/Package.swift:5]: Would it make sense to add this paragraph of yours to Package.swift?
> The CMake build system is the only one that's able to produce a working compiler. There is also a Package.swift that makes it easier to build and work with the swiftASTGen library within IDEs, but it's mainly there for editing---it won't create something that can be meaningfully executed. Most things with the new Swift parser are better implemented/tested within or on top of the swift-syntax package.

**Review by ahoppen — APPROVED** [review]
The overall build structure looks good to me. I didn’t review any of the new Swift or C++ code because IIUC that’s not the intention of this PR.

File: lib/AST/DeclContext.cpp (hunk 2 above)
Comment (DougGregor) [reply lib/AST/DeclContext.cpp:885]: No, it is not! I'll bring it back.

File: lib/ASTGen/Package.swift (hunk 3 above)
Comment (DougGregor) [reply lib/ASTGen/Package.swift:5]: Great idea!

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
TO BE CLEAR, I have no idea what the flamingo is about. @zoecarver gonna hafta answer that one. I just work here.

**Comment by zoecarver** [comment]
> TO BE CLEAR, I have no idea what the flamingo is about. @zoecarver gonna hafta answer that one. I just work here.

Just keeping things organized...

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please test macOS

**Comment by DougGregor** [comment]
Windows test failure is unrelated

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Pull request merged**